Feature request: Normalize whitespace before error output

When line numbers of a file end up changing from e.g. 99 to 101, while the line number itself is stripped, the amount of characters in the line number end up changing, producing ugly diffs like:

```diff
--- a/test-ui/ui/msg_send_id_alloc_init_different.stderr
+++ b/test-ui/ui/msg_send_id_alloc_init_different.stderr
@@ -1,14 +1,14 @@
 error[E0271]: type mismatch resolving `<Id<objc2::runtime::Object, Shared> as MaybeUnwrap>::Input == Id<NSObject, _>`
-  --> ui/msg_send_id_alloc_init_different.rs
-   |
-   |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, init] };
-   |                                          ^^^^^^^^^^^^^^^^^^^^^^^ expected struct `NSObject`, found struct `objc2::runtime::Object`
-   |
-   = note: expected struct `Id<NSObject, _>`
-              found struct `Id<objc2::runtime::Object, Shared>`
+   --> ui/msg_send_id_alloc_init_different.rs
+    |
+    |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, init] };
+    |                                          ^^^^^^^^^^^^^^^^^^^^^^^ expected struct `NSObject`, found struct `objc2::runtime::Object`
+    |
+    = note: expected struct `Id<NSObject, _>`
+               found struct `Id<objc2::runtime::Object, Shared>`
 note: required by a bound in `send_message_id`
```

It would be nice if this could have been avoided!

<details><summary><strong>More detailed example</strong></summary>
<p>

I'd like the following
```
error[E0308]: mismatched types
   --> ui/msg_send_id_invalid_receiver.rs
    |
    |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, init] };
    |                                          -------------^^^-------
    |                                          |            |
    |                                          |            expected struct `Allocated`, found struct `Id`
    |                                          arguments to this function are incorrect
    |
    = note: expected enum `Option<Allocated<_>>`
               found enum `Option<Id<objc2::runtime::Object, Shared>>`
note: associated function defined here
   --> $WORKSPACE/objc2/src/__macro_helpers.rs
    |
    |     unsafe fn send_message_id<A: MessageArguments, R: MaybeUnwrap<Input = U>>(
    |               ^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Id<objc2::runtime::Object, Shared>: MessageReceiver` is not satisfied
  --> ui/msg_send_id_invalid_receiver.rs
   |
   |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, new] };
   |                                          -------------^^^------
   |                                          |            |
   |                                          |            the trait `MessageReceiver` is not implemented for `Id<objc2::runtime::Object, Shared>`
   |                                          required by a bound introduced by this call
   |
   = help: the following other types implement trait `MessageReceiver`:
             &'a Id<T, O>
             &'a mut Id<T, objc2::rc::Owned>
   = note: required for `RetainSemantics<1>` to implement `MsgSendId<Id<objc2::runtime::Object, Shared>, Id<_, _>>`
```

to be turned into:
```
error[E0308]: mismatched types
--> ui/msg_send_id_invalid_receiver.rs
 |
 |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, init] };
 |                                          -------------^^^-------
 |                                          |            |
 |                                          |            expected struct `Allocated`, found struct `Id`
 |                                          arguments to this function are incorrect
 |
 = note: expected enum `Option<Allocated<_>>`
            found enum `Option<Id<objc2::runtime::Object, Shared>>`
note: associated function defined here
 --> $WORKSPACE/objc2/src/__macro_helpers.rs
  |
  |     unsafe fn send_message_id<A: MessageArguments, R: MaybeUnwrap<Input = U>>(
  |               ^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Id<objc2::runtime::Object, Shared>: MessageReceiver` is not satisfied
--> ui/msg_send_id_invalid_receiver.rs
 |
 |     let _: Id<Object, Shared> = unsafe { msg_send_id![obj, new] };
 |                                          -------------^^^------
 |                                          |            |
 |                                          |            the trait `MessageReceiver` is not implemented for `Id<objc2::runtime::Object, Shared>`
 |                                          required by a bound introduced by this call
 |
 = help: the following other types implement trait `MessageReceiver`:
           &'a Id<T, O>
           &'a mut Id<T, objc2::rc::Owned>
 = note: required for `RetainSemantics<1>` to implement `MsgSendId<Id<objc2::runtime::Object, Shared>, Id<_, _>>`
```

Note in particular that the first error has three spaces that you'd need to strip, while the second and third error only has two; and that these errors may appear in the same output.

</p>
</details>

I imagine the algorithm can be something like:
- When encountering `-->`, count number of spaces
- Dedent everything following it, until hitting something that can't be dedented or a new `-->`

--- a/src/normalize.rs
+++ b/src/normalize.rs
@@ -4,6 +4,7 @@ mod tests;
 
 use crate::directory::Directory;
 use crate::run::PathDependency;
+use std::cmp;
 use std::path::Path;
 
 #[derive(Copy, Clone)]
@@ -59,6 +60,7 @@ pub fn diagnostics(output: &str, context: Context) -> Variations {
         ArrowOtherCrate,
         RelativeToDir,
         LinesOutsideInputFile,
+        Unindent,
     ]
     .iter()
     .map(|normalization| apply(&output, *normalization, context))
@@ -97,6 +99,7 @@ enum Normalization {
     ArrowOtherCrate,
     RelativeToDir,
     LinesOutsideInputFile,
+    Unindent,
     // New normalization steps are to be inserted here at the end so that any
     // snapshots saved before your normalization change remain passing.
 }
@@ -120,6 +123,10 @@ fn apply(original: &str, normalization: Normalization, context: Context) -> Stri
                 normalized.push('\n');
             }
         }
+    }
+
+    if normalization >= Unindent {
+        normalized = unindent(normalized);
     }
 
     trim(normalized)
@@ -401,4 +408,122 @@ fn replace_case_insensitive(line: &str, pattern: &str, replacement: &str) -> Str
     }
 
     replaced
+}
+
+#[derive(PartialEq)]
+enum IndentedLineKind {
+    // `error`
+    // `warning`
+    Heading,
+
+    // Contains max number of spaces that can be cut based on this line.
+    // `   --> foo` = 2
+    // `    | foo` = 3
+    // `   ::: foo` = 2
+    // `10  | foo` = 1
+    Code(usize),
+
+    // `note:`
+    // `...`
+    Note,
+
+    // Contains number of leading spaces.
+    Other(usize),
+}
+
+fn unindent(diag: String) -> String {
+    let mut normalized = String::new();
+    let mut lines = diag.lines();
+
+    while let Some(line) = lines.next() {
+        normalized.push_str(line);
+        normalized.push('\n');
+
+        if indented_line_kind(line) != IndentedLineKind::Heading {
+            continue;
+        }
+
+        let mut ahead = lines.clone();
+        let next_line = match ahead.next() {
+            Some(line) => line,
+            None => continue,
+        };
+
+        if let IndentedLineKind::Code(indent) = indented_line_kind(next_line) {
+            if next_line[indent + 1..].starts_with("--> ") {
+                let mut lines_in_block = 1;
+                let mut least_indent = indent;
+                while let Some(line) = ahead.next() {
+                    match indented_line_kind(line) {
+                        IndentedLineKind::Heading => break,
+                        IndentedLineKind::Code(indent) => {
+                            lines_in_block += 1;
+                            least_indent = cmp::min(least_indent, indent);
+                        }
+                        IndentedLineKind::Note => lines_in_block += 1,
+                        IndentedLineKind::Other(spaces) => {
+                            if spaces > 10 {
+                                lines_in_block += 1;
+                            } else {
+                                break;
+                            }
+                        }
+                    }
+                }
+                for _ in 0..lines_in_block {
+                    let line = lines.next().unwrap();
+                    if let IndentedLineKind::Code(_) | IndentedLineKind::Other(_) =
+                        indented_line_kind(line)
+                    {
+                        let space = line.find(' ').unwrap();
+                        normalized.push_str(&line[..space]);
+                        normalized.push_str(&line[space + least_indent..]);
+                    } else {
+                        normalized.push_str(line);
+                    }
+                    normalized.push('\n');
+                }
+            }
+        }
+    }
+
+    normalized
+}
+
+fn indented_line_kind(line: &str) -> IndentedLineKind {
+    if let Some(heading_len) = if line.starts_with("error") {
+        Some("error".len())
+    } else if line.starts_with("warning") {
+        Some("warning".len())
+    } else {
+        None
+    } {
+        if line[heading_len..].starts_with(&[':', '['][..]) {
+            return IndentedLineKind::Heading;
+        }
+    }
+
+    if line.starts_with("note:") || line == "..." {
+        return IndentedLineKind::Note;
+    }
+
+    let is_space = |b: &u8| *b == b' ';
+    if line.starts_with("... ") {
+        let spaces = line[4..].bytes().take_while(is_space).count();
+        return IndentedLineKind::Code(spaces);
+    }
+
+    let digits = line.bytes().take_while(u8::is_ascii_digit).count();
+    let spaces = line[digits..].bytes().take_while(|b| *b == b' ').count();
+    let rest = &line[digits + spaces..];
+    if spaces > 0
+        && (rest == "|"
+            || rest.starts_with("| ")
+            || digits == 0
+                && (rest.starts_with("--> ") || rest.starts_with("::: ") || rest.starts_with("= ")))
+    {
+        return IndentedLineKind::Code(spaces - 1);
+    }
+
+    IndentedLineKind::Other(if digits == 0 { spaces } else { 0 })
 }

--- a/src/tests.rs
+++ b/src/tests.rs
@@ -62,7 +62,7 @@ error[E0277]: the trait bound `QueryParams: serde::de::Deserialize<'de>` is not 
    --> \\git\\trybuild\\test_suite\\tests\\ui\\error.rs:22:61
 " "
 error[E0277]: the trait bound `QueryParams: serde::de::Deserialize<'de>` is not satisfied
-   --> tests/ui/error.rs:22:61
+ --> tests/ui/error.rs:22:61
 "}
 
 test_normalize! {test_rust_lib
@@ -83,18 +83,18 @@ error[E0599]: no method named `quote_into_iter` found for struct `std::net::Ipv4
    | doesn't satisfy `std::net::Ipv4Addr: quote::to_tokens::ToTokens`
 " "
 error[E0599]: no method named `quote_into_iter` found for struct `std::net::Ipv4Addr` in the current scope
-  --> tests/ui/not-repeatable.rs:6:13
-   |
-6  |     let _ = quote! { #(#ip)* };
-   |             ^^^^^^^^^^^^^^^^^^ method not found in `std::net::Ipv4Addr`
-   |
-  ::: $RUST/src/libstd/net/ip.rs
-  ::: $RUST/std/src/net/ip.rs
-   |
-   | pub struct Ipv4Addr {
-   | -------------------
-   | |
-   | doesn't satisfy `std::net::Ipv4Addr: quote::to_tokens::ToTokens`
+ --> tests/ui/not-repeatable.rs:6:13
+  |
+6 |     let _ = quote! { #(#ip)* };
+  |             ^^^^^^^^^^^^^^^^^^ method not found in `std::net::Ipv4Addr`
+  |
+ ::: $RUST/src/libstd/net/ip.rs
+ ::: $RUST/std/src/net/ip.rs
+  |
+  | pub struct Ipv4Addr {
+  | -------------------
+  | |
+  | doesn't satisfy `std::net::Ipv4Addr: quote::to_tokens::ToTokens`
 "}
 
 test_normalize! {test_type_dir_backslash
@@ -111,14 +111,14 @@ error[E0277]: `*mut _` cannot be shared between threads safely
     = note: required because it appears within the type `[closure@/git/trybuild/test_suite/ui/compile-fail-3.rs:7:19: 9:6 x:&*mut _]`
 " "
 error[E0277]: `*mut _` cannot be shared between threads safely
-   --> tests/ui/compile-fail-3.rs:7:5
-    |
-7   |     thread::spawn(|| {
-    |     ^^^^^^^^^^^^^ `*mut _` cannot be shared between threads safely
-    |
-    = help: the trait `std::marker::Sync` is not implemented for `*mut _`
-    = note: required because of the requirements on the impl of `std::marker::Send` for `&*mut _`
-    = note: required because it appears within the type `[closure@$DIR/ui/compile-fail-3.rs:7:19: 9:6 x:&*mut _]`
+ --> tests/ui/compile-fail-3.rs:7:5
+  |
+7 |     thread::spawn(|| {
+  |     ^^^^^^^^^^^^^ `*mut _` cannot be shared between threads safely
+  |
+  = help: the trait `std::marker::Sync` is not implemented for `*mut _`
+  = note: required because of the requirements on the impl of `std::marker::Send` for `&*mut _`
+  = note: required because it appears within the type `[closure@$DIR/ui/compile-fail-3.rs:7:19: 9:6 x:&*mut _]`
 "}
 
 test_normalize! {test_strip_path_dependencies "
@@ -167,15 +167,15 @@ For more information about this error, try `rustc --explain E0277`.
 error: could not compile `testing` due to previous error
 " "
 error[E0277]: the trait bound `Thread: serde::de::Deserialize<'_>` is not satisfied
-    --> src/main.rs:2:36
-     |
-2    |     let _ = serde_json::from_str::<std::thread::Thread>(\"???\");
-     |                                    ^^^^^^^^^^^^^^^^^^^ the trait `serde::de::Deserialize<'_>` is not implemented for `Thread`
-     |
-    ::: $CARGO/serde_json-1.0.64/src/de.rs
-     |
-     |     T: de::Deserialize<'a>,
-     |        ------------------- required by this bound in `serde_json::from_str`
+ --> src/main.rs:2:36
+  |
+2 |     let _ = serde_json::from_str::<std::thread::Thread>(\"???\");
+  |                                    ^^^^^^^^^^^^^^^^^^^ the trait `serde::de::Deserialize<'_>` is not implemented for `Thread`
+  |
+ ::: $CARGO/serde_json-1.0.64/src/de.rs
+  |
+  |     T: de::Deserialize<'a>,
+  |        ------------------- required by this bound in `serde_json::from_str`
 "}
 
 test_normalize! {test_traits_must_be_implemented "
@@ -222,46 +222,46 @@ note: the following traits must be implemented
     = note: this error originates in the macro `anyhow` (in Nightly builds, run with -Z macro-backtrace for more info)
 " "
 error[E0599]: the method `anyhow_kind` exists for reference `&Error`, but its trait bounds were not satisfied
-   --> src/main.rs:7:13
-    |
-4   | struct Error;
-    | -------------
-    | |
-    | doesn't satisfy `Error: Into<anyhow::Error>`
-    | doesn't satisfy `Error: anyhow::private::kind::TraitKind`
-    | doesn't satisfy `Error: std::fmt::Display`
+ --> src/main.rs:7:13
+  |
+4 | struct Error;
+  | -------------
+  | |
+  | doesn't satisfy `Error: Into<anyhow::Error>`
+  | doesn't satisfy `Error: anyhow::private::kind::TraitKind`
+  | doesn't satisfy `Error: std::fmt::Display`
 ...
-7   |     let _ = anyhow!(Error);
-    |             ^^^^^^^^^^^^^^ method cannot be called on `&Error` due to unsatisfied trait bounds
-    |
-    = note: the following trait bounds were not satisfied:
-            `Error: Into<anyhow::Error>`
-            which is required by `Error: anyhow::private::kind::TraitKind`
-            `Error: std::fmt::Display`
-            which is required by `&Error: anyhow::private::kind::AdhocKind`
-            `&Error: Into<anyhow::Error>`
-            which is required by `&Error: anyhow::private::kind::TraitKind`
+7 |     let _ = anyhow!(Error);
+  |             ^^^^^^^^^^^^^^ method cannot be called on `&Error` due to unsatisfied trait bounds
+  |
+  = note: the following trait bounds were not satisfied:
+          `Error: Into<anyhow::Error>`
+          which is required by `Error: anyhow::private::kind::TraitKind`
+          `Error: std::fmt::Display`
+          which is required by `&Error: anyhow::private::kind::AdhocKind`
+          `&Error: Into<anyhow::Error>`
+          which is required by `&Error: anyhow::private::kind::TraitKind`
 note: the following traits must be implemented
-   --> $RUST/core/src/convert/mod.rs
-    |
-    | / pub trait Into<T>: Sized {
-    | |     /// Performs the conversion.
-    | |     #[stable(feature = \"rust1\", since = \"1.0.0\")]
-    | |     fn into(self) -> T;
-    | | }
-    | |_^
-    |
-   ::: $RUST/core/src/fmt/mod.rs
-    |
-    | / pub trait Display {
-    | |     /// Formats the value using the given formatter.
-    | |     ///
-    | |     /// # Examples
-...   |
-    | |     fn fmt(&self, f: &mut Formatter<'_>) -> Result;
-    | | }
-    | |_^
-    = note: this error originates in the macro `anyhow` (in Nightly builds, run with -Z macro-backtrace for more info)
+ --> $RUST/core/src/convert/mod.rs
+  |
+  | / pub trait Into<T>: Sized {
+  | |     /// Performs the conversion.
+  | |     #[stable(feature = \"rust1\", since = \"1.0.0\")]
+  | |     fn into(self) -> T;
+  | | }
+  | |_^
+  |
+ ::: $RUST/core/src/fmt/mod.rs
+  |
+  | / pub trait Display {
+  | |     /// Formats the value using the given formatter.
+  | |     ///
+  | |     /// # Examples
+... |
+  | |     fn fmt(&self, f: &mut Formatter<'_>) -> Result;
+  | | }
+  | |_^
+  = note: this error originates in the macro `anyhow` (in Nightly builds, run with -Z macro-backtrace for more info)
 "}
 
 test_normalize! {test_pyo3_url
@@ -302,20 +302,18 @@ note: required by a bound in `dropshot::Query`
 547 | pub struct Query<QueryType: DeserializeOwned + JsonSchema + Send + Sync> {
     |                                                ^^^^^^^^^^ required by this bound in `dropshot::Query`
 "
-// TODO: it would be nice to also unindent the column of `|` by one column.
-// https://github.com/dtolnay/trybuild/issues/86
 "
 error[E0277]: the trait bound `QueryParams: schemars::JsonSchema` is not satisfied
-   --> tests/fail/bad_endpoint4.rs:24:14
-    |
-24  |     _params: Query<QueryParams>,
-    |              ^^^^^^^^^^^^^^^^^^ the trait `schemars::JsonSchema` is not implemented for `QueryParams`
-    |
+  --> tests/fail/bad_endpoint4.rs:24:14
+   |
+24 |     _params: Query<QueryParams>,
+   |              ^^^^^^^^^^^^^^^^^^ the trait `schemars::JsonSchema` is not implemented for `QueryParams`
+   |
 note: required by a bound in `dropshot::Query`
-   --> src/handler.rs
-    |
-    | pub struct Query<QueryType: DeserializeOwned + JsonSchema + Send + Sync> {
-    |                                                ^^^^^^^^^^ required by this bound in `dropshot::Query`
+  --> src/handler.rs
+   |
+   | pub struct Query<QueryType: DeserializeOwned + JsonSchema + Send + Sync> {
+   |                                                ^^^^^^^^^^ required by this bound in `dropshot::Query`
 "}
 
 test_normalize! {test_uniffi_out_dir
@@ -334,14 +332,14 @@ error[E0277]: the trait bound `Arc<Counter>: FfiConverter` is not satisfied
     = note: required by `try_lift`
 " "
 error[E0277]: the trait bound `Arc<Counter>: FfiConverter` is not satisfied
-   --> $OUT_DIR[uniffi_uitests]/counter.uniffi.rs
-    |
-    |             match <std::sync::Arc<Counter> as uniffi::FfiConverter>::try_lift(ptr) {
-    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `FfiConverter` is not implemented for `Arc<Counter>`
-    |
-    = help: the following implementations were found:
-              <Arc<T> as FfiConverter>
-    = note: required by `try_lift`
+ --> $OUT_DIR[uniffi_uitests]/counter.uniffi.rs
+  |
+  |             match <std::sync::Arc<Counter> as uniffi::FfiConverter>::try_lift(ptr) {
+  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `FfiConverter` is not implemented for `Arc<Counter>`
+  |
+  = help: the following implementations were found:
+            <Arc<T> as FfiConverter>
+  = note: required by `try_lift`
 "}
 
 test_normalize! {test_proc_macro_panic

